### PR TITLE
Use -D_GLIBCXX_ASSERTIONS and other security flags

### PR DESCRIPTION
**Pull request by t-wissmann**
In the debug build, activate range checks for C++ datatypes. This flag
would have been prevented crash #1056 on fedora, that has these flags
per default. While at it, I've also added further stack protection flags
in the hope that this helps us to detect bugs earlier.

### CODE REVIEW

**Comment by t-wissmann** [comment]
The fedora flags are using `-Wp,-D_GLIBCXX_ASSERTIONS` instead of `-D_GLIBCXX_ASSERTIONS` and I don't know what the difference/relevance is. With `-D_GLIBCXX_ASSERTIONS` I can reproduce the crash of #1056 (and there's a similar story for this [project on github](https://github.com/psi4/psi4/issues/1255)). Fedora lists their [build flags here](https://fedoraproject.org/wiki/Changes/HardeningFlags28).

**Comment by sham1** [comment]
> The fedora flags are using `-Wp,-D_GLIBCXX_ASSERTIONS` instead of `-D_GLIBCXX_ASSERTIONS` and I don't know what the difference/relevance is.

`-Wp` just passes that flag to the preprocessor and bypasses the compiler.

**Comment by t-wissmann** [comment]
> `-Wp` just passes that flag to the preprocessor and bypasses the compiler. Because the compiler could change and interpret the flags by itself, it's just probably for more reliability.

Makes sense, thanks!

File: cmake/debugflags.cmake [hunk 1]
```diff
@@ -15,6 +15,14 @@ set(FLAGS
     # Disable unused parameter warnings (most cases are on purpose):
     -Wno-unused-parameter
 
+    # range checks for C++ datatypes:
+    -D_GLIBCXX_ASSERTIONS
+
+    # more stack protection:
+    -fstack-protector-strong
+    -fcf-protection
+    -fstack-clash-protection
```
Comment (blarz) [inline cmake/debugflags.cmake:24]: I think that flag is only available with gcc, that is why I put it further down in this file #1062
(But Travis seems to be asleep right now)

**Review by blarz — CHANGES_REQUESTED** [review]

File: cmake/debugflags.cmake (hunk 1 above)
Comment (t-wissmann) [reply cmake/debugflags.cmake]: Ah, I see, thanks! I've moved the flags to the GNU-section

Comment (blarz) [reply cmake/debugflags.cmake]: -fstack-protector-strong is available in both gcc and clang, only fstack-clash-protection is gcc only.
Not sure about cf-protection.

Sorry, should've been more specific

**Comment by blarz** [comment]
Should we also add FORTIFY_SOURCE?

File: cmake/debugflags.cmake (hunk 1 above)
Comment (t-wissmann) [reply cmake/debugflags.cmake]: I think it's fine if we stay safe here (not sure whether there are any c++ compilers besider clang and gcc that possibly don't support it..).

**Comment by t-wissmann** [comment]
> Should we also add FORTIFY_SOURCE?

I don't have any experience with it; if you can recommend it and if it does not hide any crashs, we can add it.

**Comment by blarz** [comment]
> > Should we also add FORTIFY_SOURCE?
> 
> I don't have any experience with it; if you can recommend it and if it does not hide any crashs, we can add it.

I don't have any firsthand experience with it either. So maybe let's stick with the flags you introduce here for now.

File: cmake/debugflags.cmake (hunk 1 above)
Comment (blarz) [reply cmake/debugflags.cmake]: Alright!

**Review by blarz — APPROVED** [review]

**Pull request merged**